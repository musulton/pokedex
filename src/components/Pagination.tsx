interface Props {
    currentPage: number
    totalCount: number
    onChangePage: (number) => void
    pageSize: number
}

const Pagination = ({
    currentPage,
    totalCount,
    onChangePage,
    pageSize
}: Props): JSX.Element => {
    const lastPage: number = Math.ceil(totalCount / pageSize)

    const previousPage = (): void => {
        if (currentPage !== 1) {
            const toPage = currentPage - 1
            onChangePage(toPage)
        }
    };

    const nextPage = (): void => {
        if (currentPage !== lastPage) {
            const toPage = currentPage + 1
            onChangePage(toPage)
        }
    };

    return (
        <div>
            <button onClick={previousPage} disabled={currentPage === 1}>
                Previous
            </button>
            &nbsp;
            <button onClick={nextPage} disabled={currentPage === lastPage}>
                Next
            </button>
        </div>
    );
};

export default Pagination;
