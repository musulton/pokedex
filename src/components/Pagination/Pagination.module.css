.paginationWrapper {
    padding: 20px 0;
    text-align: center;
}

.paginationButton {
    background-color: #ECF0F1;
    color: black;
    padding: 10px;
    width: 100px;
    border: 2px solid #2C3E50;
    margin: 0 5px;
}

.paginationButton:hover {
    background-color: #2C3E50;
    color: white;
    cursor: pointer;
    border-color: #2C3E50;
}

.paginationButton:disabled,
.paginationButton[disabled]{
     border: 1px solid #999999;
     background-color: #cccccc;
     color: #666666;
 }
