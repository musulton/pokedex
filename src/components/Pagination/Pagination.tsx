import styles from "./Pagination.module.css";
import {useMemo} from "react";

interface Props {
    currentPage: number
    totalCount: number
    onChangePage: (page: number) => void
    pageSize: number
    onChangeMaxPageLimit: (page: number) => void
    onChangeMinPageLimit: (page: number) => void
    maxPageLimit: number
    minPageLimit: number
    pageNumberLimit: number
}

const Pagination = ({
    currentPage,
    totalCount,
    onChangePage,
    pageSize,
    pageNumberLimit,
    onChangeMaxPageLimit,
    onChangeMinPageLimit,
    maxPageLimit,
    minPageLimit
}: Props): JSX.Element => {
    const lastPage: number = Math.ceil(totalCount / pageSize)
    const pages = useMemo(() => {
        const _pages = []
        for (let i = 1; i <= lastPage; i++) {
            _pages.push(i)
        }

        return _pages
    }, [lastPage]);

    const previousPage = (): void => {
        if (currentPage !== 1) {
            const toPage = currentPage - 1

            if (toPage % pageNumberLimit === 0) {
                onChangeMaxPageLimit(maxPageLimit - pageNumberLimit)
                onChangeMinPageLimit(minPageLimit - pageNumberLimit)
            }

            onChangePage(toPage)
        }
    };

    const nextPage = (): void => {
        if (currentPage !== lastPage) {
            const toPage = currentPage + 1

            if (toPage > maxPageLimit) {
                onChangeMaxPageLimit(maxPageLimit + pageNumberLimit)
                onChangeMinPageLimit(minPageLimit + pageNumberLimit)
            }

            onChangePage(toPage)
        }
    };

    const onPageNumberClick = (e: any) => {
        onChangePage(Number(e.target.id))
    }

    return (
        <div className={styles.paginationWrapper}>
            <button className={styles.paginationButton} onClick={previousPage} disabled={currentPage === 1}>
                &#8249; Previous
            </button>
            <div className={styles.pageNumberWrapper}>
                {minPageLimit >= 1 && (
                    <button
                        className={styles.dots}
                        disabled={true}
                    >
                        &hellip;
                    </button>
                )}
                {pages.map(page => {
                    if (page <= maxPageLimit && page > minPageLimit) {
                        return (
                            <button
                                className={styles.pageNumberButton}
                                key={page}
                                id={String(page)}
                                disabled={page === currentPage}
                                onClick={onPageNumberClick}
                            >
                                {page}
                            </button>
                        )
                    }
                })}
                {pages.length > maxPageLimit && (
                    <button
                        className={styles.dots}
                        disabled={true}
                    >
                        &hellip;
                    </button>
                )}
            </div>
            <button className={styles.paginationButton} onClick={nextPage} disabled={currentPage === lastPage}>
                Next &#8250;
            </button>
        </div>
    );
};

export default Pagination;
