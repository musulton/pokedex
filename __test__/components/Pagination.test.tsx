import {render, screen} from "@testing-library/react";
import {fireEvent} from "@testing-library/react";

import Pagination from "../../src/components/Pagination/Pagination";

describe("Pagination component", () => {
    it("should render Pagination correctly", () => {
        const mockProps = {
            currentPage: 1,
            totalCount: 10,
            onChangePage: jest.fn(),
            pageSize: 5
        }

        const {container} = render(<Pagination {...mockProps} />)

        expect(container.firstChild).toMatchSnapshot();
    })

    it("should call onChangePage with params when prev button is click and current page is not first page", () => {
        const mockOnChangePage = jest.fn()
        const mockProps = {
            currentPage: 2,
            totalCount: 10,
            onChangePage: mockOnChangePage,
            pageSize: 5
        }

        render(<Pagination {...mockProps} />)
        fireEvent.click(screen.getByText("‹ Previous"))

        expect(mockOnChangePage).toBeCalledWith(mockProps.currentPage - 1)
    })

    it("should not call onChangePage when prev button clicked and current page is first page", () => {
        const mockOnChangePage = jest.fn()
        const mockProps = {
            currentPage: 1,
            totalCount: 10,
            onChangePage: mockOnChangePage,
            pageSize: 5
        }

        render(<Pagination {...mockProps} />)
        fireEvent.click(screen.getByText("‹ Previous"))

        expect(mockOnChangePage).not.toBeCalled()
    })

    it("should call onChangePage with params when next button is click and current page is not last page", () => {
        const mockOnChangePage = jest.fn()
        const mockProps = {
            currentPage: 1,
            totalCount: 10,
            onChangePage: mockOnChangePage,
            pageSize: 5
        }

        render(<Pagination {...mockProps} />)
        fireEvent.click(screen.getByText("Next ›"))

        expect(mockOnChangePage).toBeCalledWith(mockProps.currentPage + 1)
    })

    it("should not call onChangePage with params when prev button is click and current page is last page", () => {
        const mockOnChangePage = jest.fn()
        const mockProps = {
            currentPage: 2,
            totalCount: 10,
            onChangePage: mockOnChangePage,
            pageSize: 5
        }

        render(<Pagination {...mockProps} />)
        fireEvent.click(screen.getByText("Next ›"))

        expect(mockOnChangePage).not.toBeCalled()
    })
})
