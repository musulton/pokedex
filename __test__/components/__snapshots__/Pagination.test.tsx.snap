// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component should render Pagination correctly 1`] = `
<div
  class="paginationWrapper"
>
  <button
    class="paginationButton"
    disabled=""
  >
    ‹ Previous
  </button>
  <div
    class="pageNumberWrapper"
  />
  <button
    class="paginationButton"
  >
    Next ›
  </button>
</div>
`;
